man/fi_mr.3.md

```diff
@@ -200,12 +200,12 @@ The following apply to memory registration.
 : When the FI_MR_LOCAL mode bit is set, applications must register all
   data buffers that will be accessed by the local hardware and provide
   a valid desc parameter into applicable data transfer operations.
-  When FI_MR_LOCAL is zero, applications are not required to register
+
+  When FI_MR_LOCAL is unset, applications are not required to register
   data buffers before using them for local operations (e.g. send and
-  receive data buffers).  The desc parameter into data transfer
-  operations will be ignored in this case, unless otherwise required
-  (e.g. se  FI_MR_HMEM).  It is recommended that applications pass in
-  NULL for desc when not required.
+  receive data buffers). Applications may optionally pass in a valid
+  desc parameter. If the desc parameter is NULL, any required local
+  memory registration will be handled by the provider.
```

Comment:
I'm going to guess that in practical use cases, this change will have no effect on existing apps, but it still technically breaks the ABI and needs to be treated as such.  This change may be better going into the 2.0 update, where version checks in the data path APIs wouldn't be needed.

A possible version check in the provider could be:
```
if (desc && (version >= FI_VERSION(1,?)) {
    /* the desc is valid */
} 
```

Providers that don't perform the version check are technically okay, since the field just continues to be ignored by them.

The comment should be updated to include what libfabric version the desc parameter is no longer treated as optional and must be NULL.

Reply:
FI_MR_HYBRID doesn't help, as it doesn't provide any additional value beyond a version check.  If the provider requires FI_MR_HYBRID, that's the same as requiring that the app request version 1.22 (or whatever version this lands in).  In either case, the provider either needs a runtime check in the data path or it only runs with updated apps (existing apps will fail).  The runtime check to support existing apps just changes to something like:
```
if (desc && ep->domain->mr_mode & FI_MR_HYBRID)
```
For the version check, it would need to be documented everywhere 'desc' is documented to be ignored that it instead must be either valid or NULL.  My guess is that is what apps are doing anyway.

Reply:
The benefit of an new bit would be apps have to opt into this new desc behavior. ABI is not broken in this case since apps can use API 1.22 and get the existing ignore desc behavior.